File: playground_cpp/src/screenshot.cpp
#include "screenshot.h"
#include "stb_image_write.h"
#include <stdexcept>
#include <iostream>
#include <vector>
#include <cstring>

namespace Screenshot {

void saveImageToPNG(VulkanContext& ctx,
                    VkImage image, VkFormat format,
                    uint32_t width, uint32_t height,
                    VkImageLayout currentLayout,
                    const std::string& outputPath) {
    VkDeviceSize imageSize = width * height * 4;

    // Create staging buffer (host-visible, host-coherent)
    VkBufferCreateInfo bufferInfo = {};
    bufferInfo.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
    bufferInfo.size = imageSize;
    bufferInfo.usage = VK_BUFFER_USAGE_TRANSFER_DST_BIT;

    VmaAllocationCreateInfo allocInfo = {};
    allocInfo.usage = VMA_MEMORY_USAGE_CPU_ONLY;
    allocInfo.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
                              VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;

    VkBuffer stagingBuffer;
    VmaAllocation stagingAllocation;

    VkResult result = vmaCreateBuffer(ctx.allocator, &bufferInfo, &allocInfo,
                                      &stagingBuffer, &stagingAllocation, nullptr);
    if (result != VK_SUCCESS) {
        throw std::runtime_error("Failed to create staging buffer for screenshot");
    }

    VkCommandBuffer cmd = ctx.beginSingleTimeCommands();

    // Transition image to TRANSFER_SRC if needed
    if (currentLayout != VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL) {
        VkImageMemoryBarrier barrier = {};
        barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
        barrier.oldLayout = currentLayout;
        barrier.newLayout = VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
        barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        barrier.image = image;
        barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
        barrier.subresourceRange.baseMipLevel = 0;
        barrier.subresourceRange.levelCount = 1;
        barrier.subresourceRange.baseArrayLayer = 0;
        barrier.subresourceRange.layerCount = 1;
        barrier.srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT |
                                VK_ACCESS_SHADER_WRITE_BIT;
        barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;

        vkCmdPipelineBarrier(cmd,
            VK_PIPELINE_STAGE_ALL_COMMANDS_BIT,
            VK_PIPELINE_STAGE_TRANSFER_BIT,
            0, 0, nullptr, 0, nullptr, 1, &barrier);
    }

    // Copy image to staging buffer
    VkBufferImageCopy region = {};
    region.bufferOffset = 0;
    region.bufferRowLength = 0;
    region.bufferImageHeight = 0;
    region.imageSubresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
    region.imageSubresource.mipLevel = 0;
    region.imageSubresource.baseArrayLayer = 0;
    region.imageSubresource.layerCount = 1;
    region.imageOffset = {0, 0, 0};
    region.imageExtent = {width, height, 1};

    vkCmdCopyImageToBuffer(cmd, image, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
                           stagingBuffer, 1, &region);

    ctx.endSingleTimeCommands(cmd);

    // Map and read pixels
    void* mapped;
    vmaMapMemory(ctx.allocator, stagingAllocation, &mapped);

    std::vector<uint8_t> pixels(imageSize);
    memcpy(pixels.data(), mapped, imageSize);

    vmaUnmapMemory(ctx.allocator, stagingAllocation);

    // Handle BGRA -> RGBA swizzle if format is BGRA
    if (format == VK_FORMAT_B8G8R8A8_UNORM || format == VK_FORMAT_B8G8R8A8_SRGB) {
        for (uint32_t i = 0; i < width * height; i++) {
            std::swap(pixels[i * 4 + 0], pixels[i * 4 + 2]);
        }
    }

    // Save as PNG
    int writeResult = stbi_write_png(outputPath.c_str(), width, height, 4,
                                     pixels.data(), width * 4);
    if (!writeResult) {
        throw std::runtime_error("Failed to write PNG: " + outputPath);
    }

    std::cout << "Saved " << width << "x" << height << " render to " << outputPath << std::endl;

    // Cleanup
    vmaDestroyBuffer(ctx.allocator, stagingBuffer, stagingAllocation);
}

} // namespace Screenshot

File: playground_rust/src/screenshot.rs
//! Screenshot capture: copy offscreen image to host-visible buffer and save as PNG.

use ash::vk;
use gpu_allocator::vulkan::{Allocation, AllocationCreateDesc, AllocationScheme, Allocator};
use gpu_allocator::MemoryLocation;
use std::path::Path;

/// A staging buffer for reading back pixel data from the GPU.
pub struct StagingBuffer {
    pub buffer: vk::Buffer,
    pub allocation: Option<Allocation>,
    pub size: u64,
}

impl StagingBuffer {
    /// Create a host-visible staging buffer large enough for width * height * 4 bytes.
    pub fn new(
        device: &ash::Device,
        allocator: &mut Allocator,
        width: u32,
        height: u32,
    ) -> Result<Self, String> {
        let size = (width * height * 4) as u64;

        let buffer_info = vk::BufferCreateInfo::default()
            .size(size)
            .usage(vk::BufferUsageFlags::TRANSFER_DST)
            .sharing_mode(vk::SharingMode::EXCLUSIVE);

        let buffer = unsafe {
            device
                .create_buffer(&buffer_info, None)
                .map_err(|e| format!("Failed to create staging buffer: {:?}", e))?
        };

        let requirements = unsafe { device.get_buffer_memory_requirements(buffer) };

        let allocation = allocator
            .allocate(&AllocationCreateDesc {
                name: "screenshot_staging",
                requirements,
                location: MemoryLocation::GpuToCpu,
                linear: true,
                allocation_scheme: AllocationScheme::GpuAllocatorManaged,
            })
            .map_err(|e| format!("Failed to allocate staging memory: {:?}", e))?;

        unsafe {
            device
                .bind_buffer_memory(buffer, allocation.memory(), allocation.offset())
                .map_err(|e| format!("Failed to bind staging buffer memory: {:?}", e))?;
        }

        Ok(Self {
            buffer,
            allocation: Some(allocation),
            size,
        })
    }

    /// Read back the pixel data from the mapped staging buffer.
    pub fn read_pixels(&self, width: u32, height: u32) -> Result<Vec<u8>, String> {
        let alloc = self
            .allocation
            .as_ref()
            .ok_or("Staging buffer has no allocation")?;

        let mapped = alloc
            .mapped_slice()
            .ok_or("Staging buffer is not mapped")?;

        let byte_count = (width * height * 4) as usize;
        if mapped.len() < byte_count {
            return Err(format!(
                "Mapped slice too small: {} < {}",
                mapped.len(),
                byte_count
            ));
        }

        Ok(mapped[..byte_count].to_vec())
    }

    /// Destroy the staging buffer, freeing GPU memory.
    pub fn destroy(&mut self, device: &ash::Device, allocator: &mut Allocator) {
        if let Some(alloc) = self.allocation.take() {
            let _ = allocator.free(alloc);
        }
        unsafe {
            device.destroy_buffer(self.buffer, None);
        }
    }
}

/// Record a command to transition an image layout.
pub fn cmd_transition_image(
    device: &ash::Device,
    cmd: vk::CommandBuffer,
    image: vk::Image,
    old_layout: vk::ImageLayout,
    new_layout: vk::ImageLayout,
    src_access: vk::AccessFlags,
    dst_access: vk::AccessFlags,
    src_stage: vk::PipelineStageFlags,
    dst_stage: vk::PipelineStageFlags,
) {
    let barrier = vk::ImageMemoryBarrier::default()
        .old_layout(old_layout)
        .new_layout(new_layout)
        .src_access_mask(src_access)
        .dst_access_mask(dst_access)
        .image(image)
        .subresource_range(
            vk::ImageSubresourceRange::default()
                .aspect_mask(vk::ImageAspectFlags::COLOR)
                .base_mip_level(0)
                .level_count(1)
                .base_array_layer(0)
                .layer_count(1),
        )
        .src_queue_family_index(vk::QUEUE_FAMILY_IGNORED)
        .dst_queue_family_index(vk::QUEUE_FAMILY_IGNORED);

    unsafe {
        device.cmd_pipeline_barrier(
            cmd,
            src_stage,
            dst_stage,
            vk::DependencyFlags::empty(),
            &[],
            &[],
            &[barrier],
        );
    }
}

/// Transition an image with custom mip level count and layer count.
pub fn cmd_transition_image_layers(
    device: &ash::Device,
    cmd: vk::CommandBuffer,
    image: vk::Image,
    old_layout: vk::ImageLayout,
    new_layout: vk::ImageLayout,
    src_access: vk::AccessFlags,
    dst_access: vk::AccessFlags,
    src_stage: vk::PipelineStageFlags,
    dst_stage: vk::PipelineStageFlags,
    level_count: u32,
    layer_count: u32,
) {
    let barrier = vk::ImageMemoryBarrier::default()
        .old_layout(old_layout)
        .new_layout(new_layout)
        .src_queue_family_index(vk::QUEUE_FAMILY_IGNORED)
        .dst_queue_family_index(vk::QUEUE_FAMILY_IGNORED)
        .image(image)
        .subresource_range(
            vk::ImageSubresourceRange::default()
                .aspect_mask(vk::ImageAspectFlags::COLOR)
                .base_mip_level(0)
                .level_count(level_count)
                .base_array_layer(0)
                .layer_count(layer_count),
        )
        .src_access_mask(src_access)
        .dst_access_mask(dst_access);

    unsafe {
        device.cmd_pipeline_barrier(
            cmd,
            src_stage,
            dst_stage,
            vk::DependencyFlags::empty(),
            &[],
            &[],
            &[barrier],
        );
    }
}

/// Record a command to copy an image to a staging buffer.
pub fn cmd_copy_image_to_buffer(
    device: &ash::Device,
    cmd: vk::CommandBuffer,
    image: vk::Image,
    buffer: vk::Buffer,
    width: u32,
    height: u32,
) {
    let region = vk::BufferImageCopy::default()
        .buffer_offset(0)
        .buffer_row_length(0)
        .buffer_image_height(0)
        .image_subresource(
            vk::ImageSubresourceLayers::default()
                .aspect_mask(vk::ImageAspectFlags::COLOR)
                .mip_level(0)
                .base_array_layer(0)
                .layer_count(1),
        )
        .image_offset(vk::Offset3D { x: 0, y: 0, z: 0 })
        .image_extent(vk::Extent3D {
            width,
            height,
            depth: 1,
        });

    unsafe {
        device.cmd_copy_image_to_buffer(
            cmd,
            image,
            vk::ImageLayout::TRANSFER_SRC_OPTIMAL,
            buffer,
            &[region],
        );
    }
}

/// Save RGBA pixel data as a PNG file.
pub fn save_png(pixels: &[u8], width: u32, height: u32, path: &Path) -> Result<(), String> {
    let img = image::RgbaImage::from_raw(width, height, pixels.to_vec())
        .ok_or("Failed to create image from pixel data")?;

    img.save(path)
        .map_err(|e| format!("Failed to save PNG to {:?}: {}", path, e))?;

    Ok(())
}
